compiler/passes/convert-uast.cpp

```diff
@@ -0,0 +1,642 @@
+        return INTENT_CONST_IN;
+      case uast::Formal::OUT:
+        return INTENT_OUT;
+      case uast::Formal::INOUT:
+        return INTENT_INOUT;
+      case uast::Formal::PARAM:
+        return INTENT_PARAM;
+      case uast::Formal::TYPE:
+        return INTENT_TYPE;
+    }
+
+    INT_FATAL("case not handled");
+    return INTENT_BLANK;
+  }
+
+  DefExpr* convertFormal(const uast::Formal* node) {
+    IntentTag intentTag = convertFormalIntent(node->intent());
+
+    Expr* typeExpr = convertExprOrNull(node->typeExpression());
+    Expr* initExpr = convertExprOrNull(node->initExpression());
+
+    Expr* varargsVariable = nullptr; // TODO: handle varargs
+
+    return buildArgDefExpr(intentTag, node->name().c_str(),
+                           typeExpr, initExpr, varargsVariable);
+  }
+
+  Expr* convertTaskVar(const uast::TaskVar* node) {
+    INT_FATAL("TODO");
+    return nullptr;
+  }
+
+  const char* tupleVariableName(const char* name) {
+    if (inTupleDecl && name[0] == '_' && name[1] == '\0')
+      return "chpl__tuple_blank";
+    else
+      return name;
+  }
+
+  Expr* convertVariable(const uast::Variable* node) {
+    auto stmts = new BlockStmt(BLOCK_SCOPELESS);
+
+    auto varSym = new VarSymbol(tupleVariableName(node->name().c_str()));
+    // Adjust the variable according to its kind
+    switch (node->kind()) {
+      case uast::Variable::VAR:
+        varSym->qual = QUAL_VAL;
+        break;
+      case uast::Variable::CONST:
+        varSym->addFlag(FLAG_CONST);
+        varSym->qual = QUAL_CONST;
+        break;
+      case uast::Variable::CONST_REF:
+        varSym->addFlag(FLAG_CONST);
+        varSym->addFlag(FLAG_REF_VAR);
+        varSym->qual = QUAL_CONST_REF;
+        break;
+      case uast::Variable::REF:
+        varSym->addFlag(FLAG_REF_VAR);
+        varSym->qual = QUAL_REF;
+        break;
+      case uast::Variable::PARAM:
+        varSym->addFlag(FLAG_PARAM);
+        varSym->qual = QUAL_PARAM;
+        break;
+      case uast::Variable::TYPE:
+        varSym->addFlag(FLAG_TYPE_VARIABLE);
+        break;
+      case uast::Variable::INDEX:
+        break;
+    }
+
+    Expr* typeExpr = convertExprOrNull(node->typeExpression());
+    Expr* initExpr = convertExprOrNull(node->initExpression());
+
+    auto defExpr = new DefExpr(varSym, initExpr, typeExpr);
+    stmts->insertAtTail(defExpr);
+
+    // Add a PRIM_END_OF_STATEMENT.
+    if (fDocs == false && inTupleDecl == false) {
+      CallExpr* end = new CallExpr(PRIM_END_OF_STATEMENT);
+      stmts->insertAtTail(end);
+    }
+
+    return stmts;
+  }
+
+  /// TypeDecls
+
+  Expr* convertEnum(const uast::Enum* node) {
+    INT_FATAL("TODO");
+    return nullptr;
+  }
+};
+
+/// Generic conversion calling the above functions ///
+Expr* Converter::convertAST(const uast::ASTNode* node) {
+  switch (node->tag()) {
+    #define CONVERT(NAME) \
+      case chpl::uast::asttags::NAME: \
+      { \
+        astlocMarker markAstLoc(node->id()); \
```

Comment:
Yes, I would agree. However the best that I could come up with was to make a macro for all of this stuff. But I didn't think that made this use clearer and I don't have another use that would benefit from pulling it out. (It's also unclear to me if there will be many more traversals of this style on the uAST or not). So I'll just propose we leave that for later.

If we could write tertiary virtually dispatched methods like in Chapel, we wouldn't need any of this... but I don't know of a satisfying way to do that in C++ in general. But that doesn't mean there isn't a good way in C++ - maybe I just haven't identified it yet. There are probably some variants on something Visitor-like that could be used; maybe with a template with the return type; maybe it could still be recursive too.